File: src/embed/fake.ts
/**
 * Deterministic offline provider for tests and CI.
 *
 * Ported from `legacy/python/src/qkb/embed/fake.py`.
 */
import { createHash } from "node:crypto";
import type { EmbeddingProvider } from "./types.js";

export class FakeProvider implements EmbeddingProvider {
  private readonly _dim: number;

  constructor(dimension = 8) {
    this._dim = dimension;
  }

  get dimension(): number {
    return this._dim;
  }

  get modelName(): string {
    return `fake-${this._dim}d`;
  }

  private vector(text: string): number[] {
    let raw = Buffer.alloc(0);
    let counter = 0;
    while (raw.length < this._dim * 4) {
      raw = Buffer.concat([raw, createHash("sha256").update(`${counter}:${text}`).digest()]);
      counter++;
    }
    const vals: number[] = [];
    for (let i = 0; i < this._dim; i++) {
      vals.push(raw.readInt32BE(i * 4));
    }
    const norm = Math.sqrt(vals.reduce((sum, v) => sum + v * v, 0)) || 1.0;
    return vals.map((v) => v / norm);
  }

  async embed(texts: string[]): Promise<number[][]> {
    return texts.map((t) => this.vector(t));
  }

  async embedQuery(query: string): Promise<number[]> {
    return this.vector(query);
  }
}

File: legacy/python/src/qkb/embed/fake.py
"""Deterministic offline provider for tests and CI."""

from __future__ import annotations

import hashlib
import math
import struct


class FakeProvider:
    def __init__(self, dimension: int = 8):
        self._dim = dimension

    @property
    def dimension(self) -> int:
        return self._dim

    @property
    def model_name(self) -> str:
        return f"fake-{self._dim}d"

    def _vector(self, text: str) -> list[float]:
        raw = b""
        counter = 0
        while len(raw) < self._dim * 4:
            raw += hashlib.sha256(f"{counter}:{text}".encode()).digest()
            counter += 1
        vals = [struct.unpack_from(">i", raw, i * 4)[0] for i in range(self._dim)]
        norm = math.sqrt(sum(v * v for v in vals)) or 1.0
        return [v / norm for v in vals]

    def embed(self, texts: list[str]) -> list[list[float]]:
        return [self._vector(t) for t in texts]

    def embed_query(self, query: str) -> list[float]:
        return self._vector(query)
